test/workspace.hpp

```diff
@@ -0,0 +1,160 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2023 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#pragma once
+
+#include <hip/hip_runtime.h>
+
+#include "get_handle.hpp"
+
+class Workspace
+{
+
+    // RAII class for hip allocations
+    class GPUBuffer
+    {
+    public:
+        GPUBuffer() = default;
+
+        explicit GPUBuffer(size_t num_bytes)
+        {
+            assert(num_bytes > 0);
+            auto s = hipMalloc(&buf_, num_bytes);
+            if(s != hipSuccess || buf_ == nullptr)
+            {
+                std::abort();
+            }
+        }
+
+        ~GPUBuffer()
+        {
+            auto s = hipFree(buf_);
+            buf_   = nullptr;
+            if(s != hipSuccess)
+            {
+                std::abort();
+            }
+        }
+
+        void* ptr() { return buf_; }
+        void* ptr() const { return buf_; }
+
+        GPUBuffer(const GPUBuffer&) = delete;
+        GPUBuffer& operator=(const GPUBuffer&) = delete;
+
+        GPUBuffer(GPUBuffer&& that) noexcept : buf_(that.buf_)
+        {
+            that.buf_ = nullptr; // take over ownership
+        }
+
+        GPUBuffer& operator=(GPUBuffer&& that) noexcept
+        {
+            std::swap(this->buf_, that.buf_);
+            return *this;
+        }
```

Comment:
This is the second place of doubling memory requirement on re-allocation.
I guess the most common case is reallocation from an empty state, but nevertheless.